### PR TITLE
Task/guardian key ceremony

### LINKED ISSUE #120: Admin button to join/continue key ceremony

### LINKED ISSUE #155: ✨ Complete Key Ceremony
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Suggestion

The key ceremony needs to be able to be completed and include the UI enhancements to show status as the ceremony runs.

### Possible Implementation

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by SteveMaier-IRT**
### Issue
Fixes #155
Fixes #120 

### Description
Key ceremony is working.
UI enhancements added for progress.
Spaces in guardian names work.
Used long polling instead of realtime communication.

### Testing
*Describe the best way to test or validate your PR.*

### CODE REVIEW

**Review by john-s-morgan — CHANGES_REQUESTED** [review]
Looks very good, and Demo was beautiful

**Review by john-s-morgan — APPROVED** [review]

File: src/electionguard-ui/ElectionGuard.UI/ViewModels/ViewKeyCeremonyViewModel.cs [hunk 1]
```diff
@@ -27,40 +44,100 @@ public partial class ViewKeyCeremonyViewModel : BaseViewModel
...
+            {
+                GuardianList.Add(new GuardianItem() { Name = item.GuardianId });
+            }
+        }
+        foreach (var guardian in GuardianList)
+        {
+            guardian.HasBackup = await _mediator!.HasBackup(guardian.Name);
+            (guardian.HasVerified, guardian.BadVerified) = await _mediator!.HasVerified(guardian.Name);
+            guardian.IsSelf = guardian.Name == UserName!;
+        }
```
[the remark this answers is not in the document]
Comment (SteveMaier-IRT) [reply src/electionguard-ui/ElectionGuard.UI/ViewModels/ViewKeyCeremonyViewModel.cs]: stopped the top loop if we are at the max guardian count

File: bindings/netstandard/ElectionGuard/ElectionGuard.ElectionSetup/KeyCeremonyMediator.cs [hunk 2]
```diff
@@ -542,25 +555,28 @@ private async Task<bool> ShouldAdminStartStep6()
...
     private async Task<bool> ShouldGuardianRunStep1()
     {
-        string keyCeremonyId = CeremonyDetails.KeyCeremonyId;
-        string guardianId = UserId;
+        var keyCeremonyId = CeremonyDetails.KeyCeremonyId;
+        var guardianId = UserId;
 
         GuardianPublicKeyService guardianService = new();
         var guardian = await guardianService.GetByIdsAsync(keyCeremonyId, guardianId);
+        var count = await guardianService.CountAsync(keyCeremonyId);
```
[the remark this answers is not in the document]
Comment (SteveMaier-IRT) [reply bindings/netstandard/ElectionGuard/ElectionGuard.ElectionSetup/KeyCeremonyMediator.cs]: pineapples

File: bindings/netstandard/ElectionGuard/ElectionGuard.ElectionSetup/KeyCeremonyMediator.cs [hunk 3]
```diff
@@ -612,7 +630,14 @@ private async Task RunStep1()
...
 
         // append to key ceremony (db)
-        await service.UpdatePublicKeyAsync(keyCeremonyId, currentGuardianUserName, publicKey);
+        try
+        {
+            await service.UpdatePublicKeyAsync(keyCeremonyId, currentGuardianUserName, publicKey);
+        }
+        catch (Exception ex)
+        {
+            Console.WriteLine(ex);
```
[the remark this answers is not in the document]
Comment (SteveMaier-IRT) [reply bindings/netstandard/ElectionGuard/ElectionGuard.ElectionSetup/KeyCeremonyMediator.cs]: Removed the call to console, put in a specific exception to catch and put in a comment about the logger

File: bindings/netstandard/ElectionGuard/ElectionGuard.ElectionSetup/Guardian.cs [hunk 4]
```diff
@@ -591,14 +601,22 @@ public void Save()
...
+        {
+            var privateGuardian = JsonSerializer.Deserialize<GuardianPrivateRecord>(data);
+            return privateGuardian != null ?
+                Guardian.FromPrivateRecord(privateGuardian, keyCeremonyId, guardianCount, quorum) :
+                null;
+        }
+        catch (Exception ex)
+        {
+            var m = ex.Message;
+        }
```
[the remark this answers is not in the document]
Comment (SteveMaier-IRT) [reply bindings/netstandard/ElectionGuard/ElectionGuard.ElectionSetup/Guardian.cs:619]: fixed

File: src/electionguard-ui/ElectionGuard.UI/ViewModels/GuardianHomeViewModel.cs [hunk 5]
```diff
@@ -15,16 +15,22 @@ public GuardianHomeViewModel(IServiceProvider serviceProvider, KeyCeremonyServic
         _tallies.Add(new Tally { Name = "Election Test Tally #1" });
         _tallies.Add(new Tally { Name = "Election Test Tally #2" });
         _tallies.Add(new Tally { Name = "Real Election Tally" });
+
+        _timer = Dispatcher.GetForCurrentThread()!.CreateTimer();
+        _timer.Interval = TimeSpan.FromSeconds(UISettings.LONG_POLLING_INTERVAL);
+        _timer.IsRepeating = true;
+        _timer.Tick += PollingTimer_Tick;
```
[the remark this answers is not in the document]
Comment (SteveMaier-IRT) [reply src/electionguard-ui/ElectionGuard.UI/ViewModels/GuardianHomeViewModel.cs:18]: good to keep in mind

File: src/electionguard-ui/ElectionGuard.UI.Lib/Models/KeyCeremonyStatesEnum.cs [hunk 6]
```diff
@@ -1,4 +1,6 @@
-﻿namespace ElectionGuard.UI.Lib.Models;
+﻿using System.ComponentModel.DataAnnotations;
```
[the remark this answers is not in the document]
Comment (SteveMaier-IRT) [reply src/electionguard-ui/ElectionGuard.UI.Lib/Models/KeyCeremonyStatesEnum.cs]: removed

File: src/electionguard-ui/ElectionGuard.UI.Lib/Extensions/EnumerableExtensions.cs [hunk 7]
```diff
@@ -9,6 +9,7 @@ public static IEnumerable<(T item, ulong index)> WithIndex<T>(this IEnumerable<T
 
     public static void Dispose(this IEnumerable<DisposableBase> source)
     {
+        if (source == null) return;
```
[the remark this answers is not in the document]
Comment (SteveMaier-IRT) [reply src/electionguard-ui/ElectionGuard.UI.Lib/Extensions/EnumerableExtensions.cs]: done

File: src/electionguard-ui/ElectionGuard.UI.Lib/Services/BaseDatabaseService.cs [hunk 8]
```diff
@@ -88,9 +93,17 @@ public async Task<List<T>> GetAllByFieldAsync(string fieldName, object fieldValu
...
+        {
+            var data = DbService.GetCollection<T>(table ?? _collection);
+            var item = await data.FindAsync<T>(UpdateFilter(filter));
+            return item.ToList();
+        }
+        catch (Exception ex)
+        {
+            var m = ex.Message;
+            throw;
+        }
```
[the remark this answers is not in the document]
Comment (SteveMaier-IRT) [reply src/electionguard-ui/ElectionGuard.UI.Lib/Services/BaseDatabaseService.cs]: fixed

**Pull request merged**